src/core/chain.rs

```diff
@@ -0,0 +1,223 @@
+use crate::core::{
+    convert::TryFromInput,
+    handler::{Handler, HandlerInput, HandlerResult},
+};
+use futures_util::future::BoxFuture;
+use std::{any::type_name, future::Future, marker::PhantomData, sync::Arc};
+
+/// A builder for Chain
+#[derive(Default)]
+pub struct ChainBuilder {
+    handlers: Vec<Box<dyn InputHandler + Sync>>,
+}
+
+impl ChainBuilder {
+    /// Adds a handler
+    ///
+    /// # Arguments
+    ///
+    /// * handler - Handler to add
+    ///
+    /// Handlers will be dispatched in the same order as they are added
+    pub fn add_handler<H, I>(&mut self, handler: H) -> &mut Self
+    where
+        H: Handler<I> + Sync + Clone + 'static,
+        I: TryFromInput + Sync + 'static,
+        <H::Future as Future>::Output: Into<HandlerResult>,
+    {
+        self.handlers.push(ConvertInputHandler::boxed(handler));
+        self
+    }
+
+    /// Creates a new Chain
+    pub fn build(self) -> Chain {
+        Chain {
+            handlers: Arc::new(self.handlers),
+        }
+    }
+}
+
+/// Updates Chain
+#[derive(Clone)]
+pub struct Chain {
+    handlers: Arc<Vec<Box<dyn InputHandler + Sync>>>,
```

Comment:
I think we can get rid of `InputHandler`, if we will use 
```suggestion
    handlers: Arc<Vec<BoxedHandler>>,
```
and `Arc::get_mut` to push handlers

Reply:
Handler is not object safe